simulator/src/output_settings.rs

```diff
@@ -0,0 +1,112 @@
+use crate::{display::SimulatorDisplay, theme::BinaryColorTheme};
+use embedded_graphics::{
+    geometry::{Point, Size},
+    pixelcolor::PixelColor,
+    DrawTarget,
+};
+
+/// Output settings.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct OutputSettings {
+    /// Pixel scale.
+    pub scale: u32,
+    /// Spacing between pixels.
+    pub pixel_spacing: u32,
+    /// Binary color theme.
+    pub theme: BinaryColorTheme,
+}
+
+impl OutputSettings {
+    /// Calculates the size of the framebuffer required to display the scaled display.
+    pub(crate) fn framebuffer_size<C>(&self, display: &SimulatorDisplay<C>) -> Size
+    where
+        C: PixelColor,
+    {
+        let width = display.size().width;
+        let height = display.size().height;
+        let output_width = width * self.scale + (width - 1) * self.pixel_spacing;
+        let output_height = height * self.scale + (height - 1) * self.pixel_spacing;
+
+        Size::new(output_width, output_height)
+    }
+
+    /// Translates a output coordinate to the corresponding display coordinate.
+    pub(crate) const fn output_to_display(&self, output_point: Point) -> Point {
+        let pitch = self.pixel_pitch() as i32;
+        Point::new(output_point.x / pitch, output_point.y / pitch)
+    }
+
+    pub(crate) const fn pixel_pitch(&self) -> u32 {
+        self.scale + self.pixel_spacing
+    }
+}
+
+impl Default for OutputSettings {
+    fn default() -> Self {
+        Self {
+            scale: 1,
+            pixel_spacing: 0,
+            theme: BinaryColorTheme::Default,
+        }
+    }
+}
+
+/// Output settings builder.
+pub struct OutputSettingsBuilder {
+    scale: Option<u32>,
+    pixel_spacing: Option<u32>,
+    theme: BinaryColorTheme,
+}
+
+impl OutputSettingsBuilder {
+    /// Creates new output settings builder.
+    pub fn new() -> Self {
+        Self {
+            scale: None,
+            pixel_spacing: None,
+            theme: BinaryColorTheme::Default,
+        }
+    }
+
+    /// Sets the pixel scale.
+    ///
+    /// A scale of `2` or higher is useful for viewing the simulator on high DPI displays.
+    pub fn scale(mut self, scale: u32) -> Self {
+        if scale == 0 {
+            panic!("scale must be >= 0");
+        }
+
+        self.scale = Some(scale);
+
+        self
+    }
+
+    /// Sets the binary color theme for the display to use.
+    pub fn theme(mut self, theme: BinaryColorTheme) -> Self {
+        self.theme = theme;
+
+        // Most binary color displays are small and individual pixels are hard to recognize
+        // on higher resolution screens. So apply some default scaling when no scaling was explicitly
+        // set.
+        self.scale.get_or_insert(3);
+        self.pixel_spacing.get_or_insert(1);
```

Comment:
Nice, I like it.

For a future PR, SDL seems to support getting pixel density with [`display_dpi`](https://rust-sdl2.github.io/rust-sdl2/sdl2/struct.VideoSubsystem.html#method.display_dpi) which could help here.

Reply:
This was already done before, but in a less documented way.

This PR changes the way the images are scaled and uses the e-g API for everything. I thought it was a good idea to do some dogfooding and have some place in e-g that uses the acceleration API. And this way the scaling is independent of SDL and the same code can be used for SDL and image output.

Reply:
Sounds good to me :)